source/rpt/compartmentalization.cc

```diff
@@ -0,0 +1,963 @@
+  : computing_timer(std::cout, TimerOutput::summary, TimerOutput::wall_times)
+  , mpi_communicator(MPI_COMM_WORLD)
+  , triangulation(mpi_communicator)
+  , cp_parameters(CPparameters)
+{}
+
+template <int dim>
+void
+Compartmentalization<dim>::generate_cylindrical_grid()
+{
+  GridOut grid_out;
+  GridGenerator::subdivided_cylinder(
+    triangulation,
+    cp_parameters.cp_param.subdivisions,
+    cp_parameters.cp_param.cylinder_radius,
+    cp_parameters.cp_param.cylinder_half_length);
+  triangulation.refine_global(cp_parameters.cp_param.initial_refinement);
+  std::ofstream output("triangulation.vtk");
+  grid_out.write_vtk(triangulation, output);
+  std::vector<double>     cell_index_output;
+  std::vector<Point<dim>> points;
+  for (const auto &cell : triangulation.active_cell_iterators())
+    {
+      // std::vector<double> center;
+      Point<3, double> center;
+      if (cell->is_locally_owned())
+        {
+          center = cell->center();
+          cells_and_indices.insert({cell->index(), cell});
+          points.push_back(center);
+          center.clear();
+        }
+    }
+
+  // output the center of cell to feed the Pyvista and extract the electric
+  // field and velocity at the cell centers
+  // from COMSOL or CFD simulation
+  std::string   filename = "cell_centers";
+  std::ofstream myfile;
+  std::string   sep = " ";
+  myfile.open(filename);
+  for (auto &point : points)
+    {
+      for (int j = 0; j < 3; j++)
+        {
+          myfile << point[j] << sep;
+        }
+      myfile << "\n";
+    }
+  myfile.close();
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::electric_field_stored_in_vector()
+  -> std::vector<std::vector<double>>
+{
+  std::vector<double> electric_field;
+  std::ifstream       fin("input_values_emw");
+  double              element;
+  while (fin >> element)
+    {
+      electric_field.push_back(element);
+    }
+
+
+  for (const auto &[cell_index, cell] : cells_and_indices)
+    {
+      std::vector<double> fill_matrix;
+      fill_matrix.push_back(cell_index);
+      fill_matrix.push_back(electric_field[cell_index]);
+      primer_matrix_index_emw.push_back(fill_matrix);
+      fill_matrix.clear();
+    }
+  return primer_matrix_index_emw;
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::electric_field_stored_in_map()
+  -> std::map<typename Triangulation<dim>::active_cell_iterator, double>
+{
+  // Read the electric field and write it in a vector
+  std::vector<double> electric_field;
+  std::ifstream       fin("input_values_emw");
+  double              element;
+  while (fin >> element)
+    {
+      electric_field.push_back(element);
+    }
+
+  // Fill the container with the electromagnetic field
+  for (const auto &[cell_index, cell] : cells_and_indices)
+    {
+      primer_map_cell_emw.insert({cell, electric_field[cell_index]});
+    }
+  return primer_map_cell_emw;
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::read_velocity_vector()
```

Comment:
I see you need to read a few files. In order to make it cleaner, I suggestion that you implement something related to the output vector in rpt_utilities.h. There's currently std::vector<double>, std::vector<std::vector<double>> and std::vector<Point<dim>>. They have a specific names (read_counts, read detectors...), but they are kinda general, there names could be modified to something more general "read_values" etc. But it would be nice if you add other file reader that you need there :)

Reply:
Here I'm pretty sure read_points() will work with the velocity vector